### PR TITLE
Fix compilation on kernels >= 5.1

### LINKED ISSUE #17: binder: Fail to compile against Linux 5.1.0
The `binder` kernel module fails to compile against Linux 5.1.0:
```
make: Entering directory '/usr/lib/modules/5.1.0-arch1-1-ARCH/build'
Makefile:205: ================= WARNING ================
Makefile:206: 'SUBDIRS' will be removed after Linux 5.3
Makefile:207: Please use 'M=' or 'KBUILD_EXTMOD' instead
Makefile:208: ==========================================
  CC [M]  /usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/ashmem/deps.o
  CC [M]  /usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/ashmem/ashmem.o
  LD [M]  /usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/ashmem/ashmem_linux.o
  Building modules, stage 2.
  MODPOST 1 modules
  CC      /usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/ashmem/ashmem_linux.mod.o
  LD [M]  /usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/ashmem/ashmem_linux.ko
make: Leaving directory '/usr/lib/modules/5.1.0-arch1-1-ARCH/build'
make: Entering directory '/usr/lib/modules/5.1.0-arch1-1-ARCH/build'
Makefile:205: ================= WARNING ================
Makefile:206: 'SUBDIRS' will be removed after Linux 5.3
Makefile:207: Please use 'M=' or 'KBUILD_EXTMOD' instead
Makefile:208: ==========================================
  CC [M]  /usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/binder/deps.o
  CC [M]  /usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/binder/binder.o
/usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/binder/binder.c:3406:11: error: initialization of 'vm_fault_t (*)(struct vm_fault *)' {aka 'unsigned int (*)(struct vm_fault *)'} from incompatible pointer type 'int (*)(struct vm_fault *)' [-Werror=incompatible-pointer-types]
  .fault = binder_vm_fault,
           ^~~~~~~~~~~~~~~
/usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/binder/binder.c:3406:11: note: (near initialization for 'binder_vm_ops.fault')
cc1: some warnings being treated as errors
make[1]: *** [scripts/Makefile.build:276: /usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/binder/binder.o] Error 1
make: *** [Makefile:1571: _module_/usr/src/packages/BUILD/anbox-modules-git/src/anbox-modules/binder] Error 2
make: Leaving directory '/usr/lib/modules/5.1.0-arch1-1-ARCH/build'
```

### PR DESCRIPTION
**Pull request by 86423355844265459587182778**
Based on torvalds/linux@e19f70aa02f34abd4c5740f761f4694e9a7c8b3d. Fixes #17.

### CODE REVIEW

**Comment by EHfive** [comment]
@morphis Could you please give a review

**Comment by SoulInfernoDE** [comment]
```
DKMS make.log for anbox-binder-1 for kernel 5.1.4-050104-generic (x86_64)
Do 23. Mai 19:52:55 CEST 2019
make -C /lib/modules/5.1.4-050104-generic/build V=0 M=$PWD
make[1]: Verzeichnis „/usr/src/linux-headers-5.1.4-050104-generic“ wird betreten
  CC [M]  /var/lib/dkms/anbox-binder/1/build/deps.o
  CC [M]  /var/lib/dkms/anbox-binder/1/build/binder.o
/var/lib/dkms/anbox-binder/1/build/binder.c:3406:11: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .fault = binder_vm_fault,
           ^~~~~~~~~~~~~~~
/var/lib/dkms/anbox-binder/1/build/binder.c:3406:11: note: (near initialization for ‘binder_vm_ops.fault’)
cc1: some warnings being treated as errors
scripts/Makefile.build:275: recipe for target '/var/lib/dkms/anbox-binder/1/build/binder.o' failed
make[2]: *** [/var/lib/dkms/anbox-binder/1/build/binder.o] Error 1
Makefile:1571: recipe for target '_module_/var/lib/dkms/anbox-binder/1/build' failed
make[1]: *** [_module_/var/lib/dkms/anbox-binder/1/build] Error 2
make[1]: Verzeichnis „/usr/src/linux-headers-5.1.4-050104-generic“ wird verlassen
Makefile:8: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by 86423355844265459587182778** [comment]
@SoulInfernoDE 
Please check if you have patched the source code before compilation. I have no issues compiling the modules on 5.1.4 after the patch.

**Comment by GrigorijST** [comment]
> @SoulInfernoDE
> Please check if you have patched the source code before compilation. I have no issues compiling the modules on 5.1.4 after the patch.

I can confirm it. Everything is good on 5.1.4 after the patch...

**Comment by SoulInfernoDE** [comment]
yes with the patch it works - is this patch beeing merged somehow sonewhere or do we  always need to apply this everytime on kernel 5.x ?

**Comment by WhyNotHugo** [comment]
@SoulInfernoDE Once this PR is merged, you won't have to apply the patch manually any more.

**Comment by dreamcat4** [comment]
@anbox please take the patch

File: binder/binder.c [hunk 1]
```diff
@@ -3391,7 +3391,9 @@ static void binder_vma_close(struct vm_area_struct *vma)
 	binder_defer_work(proc, BINDER_DEFERRED_PUT_FILES);
 }
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
```
Comment (morphis) [inline binder/binder.c]: It would be great if you can update the travis build configuration to let it build against 5.0 and 5.1 too.

**Comment by 86423355844265459587182778** [comment]
@morphis 
It looks like a retpoline compiler is needed - should we use GCC 8 for Travis?

**Comment by thebombzen** [comment]
Apparently I did a bunch of work on this, only to discover my work is a duplicate of this PR... although I did discover that apparently, the header change should start at kernel 4.17, which is when the typedef for vm_fault_t was added to the headers.

**Comment by Mhowser** [comment]
Any progress on getting this merged?

**Comment by Lumumba** [comment]
sorry for dumb question but how do i apply that patch?

**Comment by SoulInfernoDE** [comment]
if you need to apply patch you can look at:
https://github.com/archlinuxcn/repo/blob/b668ee4ea9749c77c55212032c5f064d6fb11107/archlinuxcn/anbox-git/fix_linux_5.1.x.patch

This was referenced here also. On debian/linux mint you need to sudo apt install patch. Then download the source code that needs to be patched. Put the patchfile in the root of the source code folder and use terminal/shall to apply the patch:
patch -p1 < fix_linux_5.1.x.patch
Then build the module again from the patched source.

**Comment by Lumumba** [comment]
i managed to do it but thank you

Am So., 16. Juni 2019 um 18:46 Uhr schrieb Soul Inferno DE <
notifications@github.com>:

**Comment by morphis** [comment]
@86423355844265459587182778 if we can use gcc8 on travis just for the relevant kernels that would be the way to go.

**Comment by Technically-Alexander** [comment]
Thanks @86423355844265459587182778, this tweak worked for me to install the binder on Fedora 30 running 5.1.15-300 kernel.

Hopefully the team implements it soon, and doesn't break anything.  Good work.  Appreciate it.

**Comment by 86423355844265459587182778** [comment]
@Technically-Alexander 
You are welcome. I do not know how to update the Travis build configuration to make Travis use GCC 8 just for kernels >= 5.0, so any help would be greatly appreciated.

**Comment by Hattshire** [comment]
> @Technically-Alexander
> You are welcome. I do not know how to update the Travis build configuration to make Travis use GCC 8 just for kernels >= 5.0, so any help would be greatly appreciated.

Here are some details about it:
https://docs.travis-ci.com/user/languages/cpp/#c11c11-and-beyond-and-toolchain-versioning

**Comment by Hattshire** [comment]
This problem with the travis build rises from the mitigation of _SPECTREv2_, released on _KVER=4.20_, backported to _4.19.2_, so builds to kernel versions equal or greater also **will** fail with an _outdated compiler_.

However, travis is not testing against them, sooo whatever, I guess

---
https://stackoverflow.com/questions/48089426/what-is-a-retpoline-and-how-does-it-work
https://lkml.org/lkml/2018/10/23/469

**Comment by Hattshire** [comment]
Just checked kernel's makefile:
https://github.com/torvalds/linux/blob/550d1f5bda33fa3b203d8cf8df1396825dbfd213/Makefile#L410
I wonder if it's taking `CC` as that :thinking:

**Comment by 86423355844265459587182778** [comment]
I managed to make Travis use GCC 8 just for kernels >= 5.0. Thanks @Hattshire!

**Comment by JasMich** [comment]
@morphis When will you update the Packages in your PPA at https://launchpad.net/~morphis/+archive/ubuntu/anbox-support/+packages ?

**Comment by hosford42** [comment]
I'm still getting this error.

```
/var/lib/dkms/anbox/1/build/binder/binder.c:3382:11: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .fault = binder_vm_fault,
           ^~~~~~~~~~~~~~~
/var/lib/dkms/anbox/1/build/binder/binder.c:3382:11: note: (near initialization for ‘binder_vm_ops.fault’)
cc1: some warnings being treated as errors
scripts/Makefile.build:288: recipe for target '/var/lib/dkms/anbox/1/build/binder/binder.o' failed
make[2]: *** [/var/lib/dkms/anbox/1/build/binder/binder.o] Error 1
Makefile:1655: recipe for target '_module_/var/lib/dkms/anbox/1/build/binder' failed

```

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic
```

```
$ uname -r
5.3.0-46-generic
```

```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/7/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 7.5.0-3ubuntu1~18.04' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04) 
```
Please let me know if there's other information I can provide to help.

**Comment by TheOneWithTheBraid** [comment]
@hosford42 I guess you have installed `linux-headers` and are building from latest upstream?

**Comment by TheOneWithTheBraid** [comment]
@hosford42 Oh, sorry, maybe #36 might fix this?

**Comment by hosford42** [comment]
That does look like it will address my error. I was attempting to follow the instructions in https://github.com/anbox/anbox/blob/master/docs/install.md, which indicate that anbox-modules should be installed from `ppa:morphis/anbox-support`.

**Pull request merged**